### ISSUE TITLE
SSR support in @fluentui/react-components

### ISSUE THREAD

**Issue by layershifter** [microsoft/fluentui#17893]
This epic tracks support of SSR in `@fluentui/react-components`.

Active issues:
- SSR tests #22788
- #22174

Solved issues:
- @fluentui/react-image: throws in SSR #17889
- @fluentui/react-tooltip: warning about useLayoutEffect in SSR #17891
- @fluentui/react-menu: throws in SSR #17899
- @fluentui/make-styles: createDOMRenderer() throws an error in SSR #17902
- @fluentui/react-utilities: useId() is not SSR safe #17892
- @fluentui/react-tooltip: throws an error during SSR #17901
- makeStyles: rendering styles on server & hydration #18039

**Comment by thekeviv** [comment]
@miroslavstastny @layershifter 
I am using FluentUI 8.22.3 in a nextjs project with nextjs version 11.0.1.

I get the useLayoutEffect warning for many more components I use. I'm not sure if the tasks completed above fixes issues with the components I am seeing them in as well. 

```
Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
    at LayerBase (\node_modules\@fluentui\react\lib-commonjs\components\Layer\Layer.base.js:14:20)
    at StyledLayerBase (\node_modules\@fluentui\utilities\lib-commonjs\styled.js:13:28)
    at Callout (\node_modules\@fluentui\react\lib-commonjs\components\Callout\Callout.js:9:25)
    at ContextualMenuInternal (\node_modules\@fluentui\react\lib-commonjs\components\ContextualMenu\ContextualMenu.base.js:152:28)
    at ContextualMenuBase (\node_modules\@fluentui\react\lib-commonjs\components\ContextualMenu\ContextualMenu.base.js:127:25)
    at StyledContextualMenuBase (\node_modules\@fluentui\utilities\lib-commonjs\styled.js:13:28)
    at div
    at Stack (\node_modules\@fluentui\foundation-legacy\lib-commonjs\createComponent.js:34:70)
    at div
    at Stack (\node_modules\@fluentui\foundation-legacy\lib-commonjs\createComponent.js:34:70)
    at div
    at Stack (\node_modules\@fluentui\foundation-legacy\lib-commonjs\createComponent.js:34:70)
    at div
    at div
    at div
    at Sticky (\node_modules\@fluentui\react\lib-commonjs\components\Sticky\Sticky.js:13:28)
    at div
    at div
    at ScrollablePaneBase (\node_modules\@fluentui\react\lib-commonjs\components\ScrollablePane\ScrollablePane.base.js:12:28)
    at StyledScrollablePaneBase (\node_modules\@fluentui\utilities\lib-commonjs\styled.js:13:28)
    at Layout (C:\Users\vivek\source\repos\MagicIT\avs-windows-portal\frontend\portal-web\.next\server\pages\_app.js:171:64)
    at div
    at ThemeProvider (\node_modules\@fluentui\react\lib-commonjs\utilities\ThemeProvider\ThemeProvider.js:13:58)
    at MyApp (\.next\server\pages\_app.js:3289:3)
    at AppContainer (\node_modules\next\dist\next-server\server\render.js:27:952)

```

If it's relevant, here's my _document.tsx file:

```
import Document, {
  DocumentContext,
  Head,
  Html,
  Main,
  NextScript,
} from "next/document";
// Fluent UI React 8+
import { Stylesheet, InjectionMode, resetIds } from "@fluentui/react";
const stylesheet = Stylesheet.getInstance();

// Set the config.
stylesheet.setConfig({
  injectionMode: InjectionMode.none,
  namespace: "server",
});

// Now set up the document, and just reset the stylesheet.
export default class MyDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    stylesheet.reset();
    resetIds();
    const initialProps = await Document.getInitialProps(ctx);

    return {
      ...initialProps,
      styles: [
        initialProps.styles,
        <style
          key="fluentui-css"
          dangerouslySetInnerHTML={{ __html: stylesheet.getRules(true) }}
        />,
      ],
    };
  }

  render(): JSX.Element {
    return (
      <Html lang="pt-BR">
        <Head>
          <meta charSet="utf-8" />
        </Head>

        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

```

**Comment by layershifter** [comment]
@thekeviv this issue related to `@fluentui/react-components` aka `vNext` (v8 is `@fluentui/react`). Can you please create a separate issue for your problem?

**Comment by layershifter** [comment]
I implemented tests in #23457, the only remaining issue is #22174, but it's a regular bug.
Closing it.
